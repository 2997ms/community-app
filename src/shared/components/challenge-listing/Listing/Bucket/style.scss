@import "~styles/tc-styles";

.bucket {
  background: $tc-gray-neutral-light;
  margin: 20px 0;
}

.view-more {
  width: 100%;
  background: $tc-gray-neutral-light;
  border-top: 1px solid $tc-gray-10;
  border-bottom: 1px solid $tc-gray-10;
  border-left: 0;
  border-right: 0;
  color: $tc-gray-50;
  font-weight: 300;
  text-transform: none;
  border-radius: 0;
  outline: none;
  font-size: 13px;
  padding: 12px 0;
}
