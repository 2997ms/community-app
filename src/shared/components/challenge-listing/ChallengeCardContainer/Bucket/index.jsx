/**
 * A single bucket of challenges.
 */

import _ from 'lodash';
import PT from 'prop-types';
import React from 'react';
import sort from 'utils/challenge-listing/sort';
import SortingSelectBar from 'components/SortingSelectBar';
import { getFilterFunction } from 'utils/challenge-listing/filter';
import ChallengeCard from '../../ChallengeCard';
import './style.scss';

const COLLAPSED_SIZE = 10;

export default function Bucket({ bucket, challenges, expanded }) {
  const filter = getFilterFunction(bucket.filter);

  let expandable = false;
  const filteredChallenges = [];
  for (let i = 0; i < challenges.length; i += 1) {
    if (filter(challenges[i])) filteredChallenges.push(challenges[i]);
    if (!expanded && filteredChallenges.length >= COLLAPSED_SIZE) {
      expandable = true;
      break;
    }
  }

  if (!filteredChallenges.length) return null;

  const cards = filteredChallenges.map(item => (
    <ChallengeCard
      challenge={item}
      onTechTagClicked={_.noop}
      key={item.id}
    />
  ));

  return (
    <div styleName="bucket">
      <SortingSelectBar
        options={bucket.sorts.map(item => sort[item].name)}
        title={bucket.name}
        value={sort[bucket.sorts[0]].name}
      />
      {cards}
      {
        expandable ? (
          <button styleName="view-more">View more challenges</button>
        ) : null
      }
    </div>
  );
}

Bucket.defaultProps = {
  expanded: false,
};

Bucket.propTypes = {
  bucket: PT.shape().isRequired,
  expanded: PT.bool,
  challenges: PT.arrayOf(PT.shape()).isRequired,
};
