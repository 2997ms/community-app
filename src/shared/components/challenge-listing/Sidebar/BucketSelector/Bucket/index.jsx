/**
 * Regular sidebar row.
 */

import _ from 'lodash';
// import { challenge as challengeUtils } from 'topcoder-react-lib';
import PT from 'prop-types';
import { connect } from 'react-redux';
import React from 'react';
import { BUCKET_NAME } from 'utils/challenge-listing/buckets';
import './style.scss';

// const Filter = challengeUtils.filter;

function Bucket({
  active,
  bucket,
  // challenges,
  // disabled,
  onClick,
  // allActiveChallengesLoaded,
  meta,
}) {
  // let countEl;
  // if (!bucket.hideCount && !disabled) {
  // const filter = Filter.getFilterFunction(bucket.filter);
  // const clonedChallenges = _.clone(challenges);
  // const filteredChallenges = [];
  // for (let i = 0; i < clonedChallenges.length; i += 1) {
  // if (filter(clonedChallenges[i])) {
  // filteredChallenges.push(clonedChallenges[i]);
  // }
  // }
  let count = 0;
  // if (allActiveChallengesLoaded) {
  //   count = challenges.filter(filter).length;
  // } else {
  switch (bucket) {
    case 'all':
      count = meta.allChallengesCount;
      break;
    case 'my':
      count = meta.myChallengesCount;
      break;
    case 'openForRegistration':
      count = meta.openChallengesCount;
      break;
    case 'ongoing':
      count = meta.ongoingChallengesCount;
      break;
    default:
  // }
  }
  const countEl = (
    <span styleName="right">
      {count}
    </span>
  );
  // }

  // const error = Boolean(bucket.error) && (
  //   <div styleName="errorMsg">
  //     {bucket.error}
  //   </div>
  // );

  if (active) {
    return (
      <div styleName="active bucket">
        {BUCKET_NAME[bucket]}
        {countEl}
        {/* {error} */}
      </div>
    );
  }

  return (
    <div
      // onClick={disabled ? _.noop : onClick}
      onClick={onClick}
      onKeyPress={e => (e.key === 'Enter' ? onClick() : null)}
      role="button"
      styleName="bucket"
      tabIndex={0}
    >
      {BUCKET_NAME[bucket]}
      {countEl}
      {/* {error} */}
    </div>
  );
}

Bucket.defaultProps = {
  active: false,
  // disabled: false,
  onClick: _.noop,
  meta: {},
};

Bucket.propTypes = {
  active: PT.bool,
  bucket: PT.string.isRequired,
  // bucket: PT.shape({
  //   hideCount: PT.bool,
  //   name: PT.string.isRequired,
  //   error: PT.string,
  //   filter: PT.any,
  // }).isRequired,
  // // challenges: PT.arrayOf(PT.shape).isRequired,
  // disabled: PT.bool,
  onClick: PT.func,
  meta: PT.shape(),
  // allActiveChallengesLoaded: PT.bool.isRequired,
};

const mapStateToProps = (state) => {
  const cl = state.challengeListing;
  return {
    allActiveChallengesLoaded: cl.allActiveChallengesLoaded,
    meta: cl.meta,
  };
};

const BucketContainer = connect(
  mapStateToProps,
)(Bucket);

export default BucketContainer;
