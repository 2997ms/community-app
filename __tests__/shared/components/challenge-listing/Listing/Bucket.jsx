import React from 'react';
import _ from 'lodash';
import Renderer from 'react-test-renderer/shallow';
import TU from 'react-dom/test-utils';
import Bucket from 'components/challenge-listing/Listing/Bucket';

const expand = jest.fn();
const loadMore = jest.fn();
const setFilterState = jest.fn();
const setSort = jest.fn();

const mockDatas = [{
  bucket: {
    sorts: ['current-phase'],
    filter: {
      status: 'ac',
    },
  },
  expanded: true,
  expand,
  challenges: [
    {
      id: '1',
      status: 'b',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '2',
      status: 'a',
      technologies: 'a, b, c',
      allPhases: [{
        phaseType: 'Registration',
        phaseStatus: 'Open',
      }],
      currentPhases: ['Registration'],
      track: 'DEVELOP',
      subTrack: 'CODE',
      events: [{ eventName: 'Submit' }],
      prizes: [1200, 600],
      totalPrize: 1800,
    }],
  loading: false,
  loadMore,
  setFilterState,
  setSort,
  sort: '',
}, {
  bucket: {
    sorts: ['current-phase'],
    filter: {
      status: 'ac',
    },
  },
  expanded: false,
  expand,
  challenges: [
    {
      id: '1',
      status: 'b',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '2',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '3',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '4',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '5',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '6',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '7',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '8',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '9',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '10',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '11',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
  ],
  loading: true,
  loadMore,
  setFilterState,
  setSort,
  sort: '',
}, {
  bucket: {
    sorts: ['current-phase'],
    filter: {
      status: 'e',
    },
  },
  expanded: false,
  expand,
  challenges: [
  ],
  loading: true,
  setFilterState,
  setSort,
  sort: '',
}];

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();

  _.forEach(mockDatas, (data) => {
    renderer.render((
      <Bucket {...data} />
    ));
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});

class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    return <Bucket {...this.props} />;
  }
}

const instance = TU.renderIntoDocument((<Wrapper {...mockDatas[0]} />));

test('setFilterState', () => {
  const matches = TU.findAllInRenderedTree(instance, item =>
    item && item.className && item.className.match('technology'));
  expect(matches).toHaveLength(3);
  TU.Simulate.click(matches[0]);
  expect(setFilterState).toHaveBeenCalledTimes(1);
  TU.Simulate.click(matches[1]);
  expect(setFilterState).toHaveBeenCalledTimes(2);
  TU.Simulate.click(matches[2]);
  expect(setFilterState).toHaveBeenCalledTimes(3);
});

