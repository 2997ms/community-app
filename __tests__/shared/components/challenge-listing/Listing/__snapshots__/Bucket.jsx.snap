// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<div
  className="src-shared-components-challenge-listing-Listing-Bucket-___style__bucket___3jwPe"
>
  <SortingSelectBar
    onSelect={[Function]}
    options={
      Array [
        Object {
          "label": "Current phase",
          "value": "current-phase",
        },
      ]
    }
    title=""
    value={
      Object {
        "label": "Current phase",
        "value": "current-phase",
      }
    }
  />
  <ChallengeCard
    challenge={
      Object {
        "allPhases": Array [
          Object {
            "phaseStatus": "Open",
            "phaseType": "Registration",
          },
        ],
        "currentPhases": Array [
          "Registration",
        ],
        "events": Array [
          Object {
            "eventName": "Submit",
          },
        ],
        "id": "2",
        "isDataScience": false,
        "prize": Array [
          1200,
          600,
        ],
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "subTrack": "CODE",
        "technologies": "a, b, c",
        "totalPrize": 1800,
        "track": "DEVELOP",
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <Waypoint
    bottomOffset="0px"
    fireOnRapidScroll={true}
    horizontal={false}
    onEnter={[Function]}
    onLeave={[Function]}
    onPositionChange={[Function]}
    topOffset="0px"
  />
</div>
`;

exports[`Matches shallow shapshot 2`] = `
<div
  className="src-shared-components-challenge-listing-Listing-Bucket-___style__bucket___3jwPe"
>
  <SortingSelectBar
    onSelect={[Function]}
    options={
      Array [
        Object {
          "label": "Current phase",
          "value": "current-phase",
        },
      ]
    }
    title=""
    value={
      Object {
        "label": "Current phase",
        "value": "current-phase",
      }
    }
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "6",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "11",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "3",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "4",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "5",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "2",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "7",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "8",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "9",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCard
    challenge={
      Object {
        "id": "10",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      }
    }
    onTechTagClicked={[Function]}
    openChallengesInNewTabs={false}
    sampleWinnerProfile={undefined}
  />
  <ChallengeCardPlaceholder
    id={0}
  />
  <ChallengeCardPlaceholder
    id={1}
  />
  <ChallengeCardPlaceholder
    id={2}
  />
  <ChallengeCardPlaceholder
    id={3}
  />
  <ChallengeCardPlaceholder
    id={4}
  />
  <ChallengeCardPlaceholder
    id={5}
  />
  <ChallengeCardPlaceholder
    id={6}
  />
  <ChallengeCardPlaceholder
    id={7}
  />
  <button
    className="src-shared-components-challenge-listing-Listing-Bucket-___style__view-more___2UekL"
    onClick={[Function]}
  >
    View more challenges
  </button>
</div>
`;

exports[`Matches shallow shapshot 3`] = `null`;
